### PR TITLE
fix(vault-agent): wait for network to be online before starting

### PR DESCRIPTION
**Pull request by jfroche**
Network needs to be up to access the vault server.

### CODE REVIEW

**Comment by Mic92** [comment]
What issue are you facing that you are trying to fix with network-online.target? Usually this target gives a false sense and can be easily mis-configured in my experience: https://systemd.io/NETWORK_ONLINE/

If vault-agent fails to start because of not reaching vault server, I would rather configure systemd to continuously restart it.

**Comment by jfroche** [comment]
We use vault-agent auto auth with the AWS auth method. vault-agent need to be online to contact AWS to get the credentials before contacting the vault server.
Here is the log of a vault-agent service that started before the network was online:

```
Jul 05 14:36:33 app systemd[1]: Starting SSH Daemon...
Jul 05 14:36:33 app systemd[1]: Started vault-agent-default.service.
Jul 05 14:36:33 app systemd[1]: Started SSH Daemon.
Jul 05 14:36:33 app sshd[3868797]: Server listening on 0.0.0.0 port 22.
Jul 05 14:36:33 app sshd[3868797]: Server listening on :: port 22.
Jul 05 14:36:33 app nginx-pre-start[3868791]: nginx: the configuration file /nix/store/w57adwiaw60dcha3sbj0wc3zyng02wr4-nginx.conf syntax is ok
Jul 05 14:36:33 app nginx-pre-start[3868791]: nginx: configuration file /nix/store/w57adwiaw60dcha3sbj0wc3zyng02wr4-nginx.conf test is successful
Jul 05 14:36:33 app systemd[1]: Started Nginx Web Server.
Jul 05 14:36:33 app systemd[1]: dhcpcd.service: Deactivated successfully.
Jul 05 14:36:33 app systemd[1]: Stopped DHCP Client.
Jul 05 14:36:33 app systemd[1]: dhcpcd.service: Consumed 6.944s CPU time, read 536.0K from disk, written 640.0K to disk, received 14.9M IP traffic, sent 77.7K IP traffic.
Jul 05 14:36:33 app systemd[1]: Starting DHCP Client...
Jul 05 14:36:33 app dhcpcd[3868816]: dhcpcd-9.4.1 starting
Jul 05 14:36:33 app dhcpcd[3868819]: dev: loaded udev
Jul 05 14:36:33 app systemd[1508939]: Created slice User Core Session Slice.
Jul 05 14:36:33 app dhcpcd[3868819]: DUID <REDACTED>
Jul 05 14:36:33 app systemd[1508939]: Starting D-Bus User Message Bus...
Jul 05 14:36:33 app systemd[1508939]: Started D-Bus User Message Bus.
Jul 05 14:36:33 app dhcpcd[3868819]: ens5: IAID <REDACTED>
Jul 05 14:36:33 app dhcpcd[3868819]: ens5: adding address <REDACTED>
Jul 05 14:36:34 app dhcpcd[3868819]: ens5: soliciting an IPv6 router
Jul 05 14:36:34 app dhcpcd[3868819]: ens5: rebinding lease of <REDACTED>
Jul 05 14:36:34 app dhcpcd[3868819]: ens5: probing address <REDACTED>/24
Jul 05 14:36:34 app vault[3868794]: Error creating aws auth method: failed to retrieve credentials from credential chain: NoCredentialProviders: no valid providers in chain. Deprecated.
Jul 05 14:36:34 app vault[3868794]:         For verbose messaging see aws.Config.CredentialsChainVerboseErrors
Jul 05 14:36:34 app systemd[1]: vault-agent-default.service: Main process exited, code=exited, status=1/FAILURE
Jul 05 14:36:34 app systemd[1]: vault-agent-default.service: Failed with result 'exit-code'.
Jul 05 14:36:34 app systemd[1]: vault-agent-default.service: Consumed 184ms CPU time, received 0B IP traffic, sent 340B IP traffic.
Jul 05 14:36:37 app systemd[1]: vault-agent-default.service: Scheduled restart job, restart counter is at 1.
Jul 05 14:36:37 app systemd[1]: Stopped vault-agent-default.service.
Jul 05 14:36:37 app systemd[1]: vault-agent-default.service: Consumed 184ms CPU time, received 0B IP traffic, sent 340B IP traffic.
Jul 05 14:36:37 app systemd[1]: Started vault-agent-default.service.
Jul 05 14:36:37 app vault[3868852]: Error creating aws auth method: failed to retrieve credentials from credential chain: NoCredentialProviders: no valid providers in chain. Deprecated.
Jul 05 14:36:37 app vault[3868852]:         For verbose messaging see aws.Config.CredentialsChainVerboseErrors
Jul 05 14:36:37 app systemd[1]: vault-agent-default.service: Main process exited, code=exited, status=1/FAILURE
Jul 05 14:36:37 app systemd[1]: vault-agent-default.service: Failed with result 'exit-code'.
Jul 05 14:36:37 app systemd[1]: vault-agent-default.service: Consumed 146ms CPU time, received 0B IP traffic, sent 340B IP traffic.
Jul 05 14:36:39 app dhcpcd[3868819]: ens5: leased <REDACTED> for 3600 seconds
Jul 05 14:36:39 app dhcpcd[3868819]: ens5: adding route to <REDACTED>/24
Jul 05 14:36:39 app dhcpcd[3868819]: ens5: adding default route via <REDACTED>
```

Note that `network-online` is also used in the official vault package released by Hashicorp: https://github.com/hashicorp/vault/blob/c6ef0800a00f3634b869e1152cfbf5581753633f/.release/linux/package/usr/lib/systemd/system/vault.service#L4-L5

Shall we do both Restart=always and After=network-online ?

**Comment by zimbatm** [comment]
We also disabled the online check in SrvOS because it's broken:
https://github.com/numtide/srvos/blob/d3b73248c271904f06aee8d0589024ceaacf497e/nixos/common/networking.nix#L14-L17

The best is to restart the service until it starts working.

File: nix/modules/vault-secrets.nix [hunk 1]
```diff
@@ -175,12 +175,15 @@ in
               Before=${service}.service
               BindsTo=${service}.service
               StopPropagatedFrom=${service}.service
+              After=network-online.target
```
Comment (Mic92) [inline nix/modules/vault-secrets.nix:178]: This actually shouldn't depend on network. It should just indefinitely block on systemd-vaultd's socket.

File: nix/modules/vault-secrets.nix [hunk 2]
```diff
@@ -175,12 +175,15 @@ in
...
 
               [Service]
               Type=oneshot
               ExecStart=${pkgs.coreutils}/bin/true
               RemainAfterExit=true
               LoadCredential=${service}.service.EnvironmentFile:/run/systemd-vaultd/sock
+              Restart=on-failure
+              RestartSec=3
```
Comment (Mic92) [inline nix/modules/vault-secrets.nix:186]: If this fails once, would this not break the original service in any case?

**Comment by jfroche** [comment]
Agree, I moved that to retrying indefinitely instead

**Pull request closed without merging**